controllers/namespace/namespace_controller.go

```diff
@@ -0,0 +1,219 @@
+package namespace
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/red-hat-storage/ocs-operator/controllers/util"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// once is a pointer to the sync.Once type
+var once sync.Once
+
+// NamespaceReconciler reconciles the namespaces starting with openshift-*
+// nolint:revive
+type NamespaceReconciler struct {
+	client.Client
+	ctx    context.Context
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=list;delete
+// +kubebuilder:rbac:groups=core,resources=namespaces,verbs=get;list;watch;update;patch
+
+// Reconcile reads that state of the cluster for a Namespace object and makes changes based on the state read
+// i.e,the annotations of the namespace
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *NamespaceReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	prevLogger := r.Log
+	defer func() { r.Log = prevLogger }()
+	r.Log = r.Log.WithValues("Request.Name", request.Name)
+	r.ctx = ctx
+
+	r.Log.Info("Reconciling Namespace.")
+
+	var err error
+	updated := false
+
+	once.Do(func() {
+		updated, err = r.annotateOpenshiftNamespaces()
+		if err != nil {
+			return
+		}
+		if updated {
+			err = r.restartCSIAddonsPod()
+		}
+	})
+
+	if err != nil {
+		return reconcile.Result{}, err
+	}
```

Comment:
This can be moved to the end of the Reconcile.

Reply:
We need it here, so that in case there is an error in `annotateOpenshiftNamespaces` func call, we should error out the reconcile and not proceed further.

Reply:
it will get reconciled again and you will add the annotation and restart it?

Reply:
Understood, updated the code